File: diagnose_db.py
#!/usr/bin/env python3
"""
Script de diagnóstico avanzado para PostgreSQL de SINFIMAC
"""
import psycopg2
from psycopg2 import sql
import sys
import re
import os

# Lista extendida de bases de datos a probar
DATABASES = [
    # Formato Supabase estándar: projectref$default
    {'database': 'postgres'},
    {'database': 'default$default'},
    # Nombres comunes
    {'database': 'supabase'},
    {'database': 'app'},
    {'database': 'sfmac'},
    {'database': 'db'},
    {'database': 'prod'},
    {'database': 'production'},
    {'database': 'main'},
    {'database': 'application'},
]

DB_CONFIG_BASE = {
    'host': '87.99.137.96',
    'port': 5432,
    'user': 'postgres',
    'password': 'CorpFlowSFMAC_DB_2026',
    'connect_timeout': 10
}

def try_connect(db_name, extra_options=None):
    """Intentar conectar a una base de datos específica"""
    config = {**DB_CONFIG_BASE, 'database': db_name}
    if extra_options:
        config.update(extra_options)
    try:
        conn = psycopg2.connect(**config)
        conn.autocommit = True
        return conn
    except Exception as e:
        return None, str(e)

def main():
    print('🔍 Buscando base de datos válida...\n')
    
    conn = None
    db_name = None
    last_error = None
    
    for db in DATABASES:
        db_name = db['database']
        print(f'Probando: {db_name}...', end=' ')
        conn, error = try_connect(db_name)
        if conn:
            print(f'✅')
            break
        else:
            print(f'❌ ({error[:50]}...)')
            last_error = error
    
    # Si no funcionó, probar con opciones adicionales
    if not conn:
        print('\n🔄 Probando con opciones adicionales...')
        
        # Probar con sslmode
        for db in DATABASES:
            db_name = db['database']
            print(f'Probando con sslmode=require: {db_name}...', end=' ')
            conn, error = try_connect(db_name, {'sslmode': 'require'})
            if conn:
                print(f'✅')
                break
            else:
                print(f'❌')
        
        # Probar con sslmode=disable
        if not conn:
            for db in DATABASES:
                db_name = db['database']
                print(f'Probando con sslmode=disable: {db_name}...', end=' ')
                conn, error = try_connect(db_name, {'sslmode': 'disable'})
                if conn:
                    print(f'✅')
                    break
                else:
                    print(f'❌')
    
    if not conn:
        print(f'\n❌ No se pudo conectar a ninguna base de datos')
        print(f'Último error: {last_error}')
        return

    print(f'\n✅ Conectado exitosamente a: {db_name}\n')
    cursor = conn.cursor()

    # 1. Listar TODAS las tablas públicas
    print('📊 TODAS LAS TABLAS PÚBLICAS\n')
    cursor.execute("""
        SELECT table_name 
        FROM information_schema.tables 
        WHERE table_schema = 'public'
        ORDER BY table_name;
    """)
    tables = [row[0] for row in cursor.fetchall()]
    print('Tablas:', ', '.join(tables) if tables else 'Ninguna')

    # 2. Buscar tablas relacionadas con tickets
    ticket_tables = [t for t in tables if 'ticket' in t.lower() or 'cost' in t.lower()]
    if ticket_tables:
        print(f'\n📋 Tablas relacionadas con tickets: {", ".join(ticket_tables)}')
        
        for table in ticket_tables:
            print(f'\n  📋 Estructura de {table}:')
            try:
                cursor.execute(f"""
                    SELECT column_name, data_type, is_nullable 
                    FROM information_schema.columns 
                    WHERE table_name = '{table}' 
                    ORDER BY ordinal_position;
                """)
                for col in cursor.fetchall():
                    nullable = '' if col[2] == 'YES' else 'NOT NULL'
                    print(f'    - {col[0]}: {col[1]} {nullable}')
                
                # Índices
                print(f'\n  📇 Índices en {table}:')
                cursor.execute(f"""
                    SELECT indexname, indexdef 
                    FROM pg_indexes 
                    WHERE tablename = '{table}';
                """)
                indexes = cursor.fetchall()
                if indexes:
                    for idx in indexes:
                        print(f'    - {idx[0]}')
                else:
                    print(f'    (ninguno)')
                
                # Conteo
                cursor.execute(f'SELECT COUNT(*) FROM {table};')
                count = cursor.fetchone()[0]
                print(f'\n  📊 Registros en {table}: {count}')
                
            except Exception as e:
                print(f'    Error: {e}')

    # 3. Buscar funciones RPC
    print('\n🔧 Funciones RPC disponibles:')
    cursor.execute("""
        SELECT proname 
        FROM pg_proc 
        WHERE pronamespace = (SELECT oid FROM pg_namespace WHERE nspname = 'public')
        ORDER BY proname;
    """)
    funcs = [row[0] for row in cursor.fetchall()]
    print('Funciones:', ', '.join(funcs) if funcs else 'Ninguna')
    
    # Filtrar las de tickets
    ticket_funcs = [f for f in funcs if 'ticket' in f.lower()]
    if ticket_funcs:
        print(f'\n🔧 Funciones relacionadas con tickets: {", ".join(ticket_funcs)}')
        for func_name in ticket_funcs:
            cursor.execute(f"""
                SELECT prosrc FROM pg_proc WHERE proname = '{func_name}'
            """)
            result = cursor.fetchone()
            if result:
                src = result[0]
                print(f'\n  - {func_name}:')
                if src and len(src) > 200:
                    print(f'    {src[:200]}...')
                else:
                    print(f'    {src}')

    cursor.close()
    conn.close()
    print('\n\n✅ Diagnóstico completado')

if __name__ == '__main__':
    main()

File: diagnose-db.js
const { Client } = require('pg');

const client = new Client({
  host: '87.99.137.96',
  port: 5432,
  database: 'postgres',
  user: 'postgres',
  password: 'CorpFlowSFMAC_DB_2026',
});

async function main() {
  console.log('🔌 Conectando a PostgreSQL...');
  await client.connect();
  console.log('✅ Conectado\n');

  // 1. Diagnosticar tablas existentes
  console.log('📊 DIAGNÓSTICO DE TABLAS\n');
  
  // Ver tablas relacionadas con tickets
  const tablesResult = await client.query(`
    SELECT table_name 
    FROM information_schema.tables 
    WHERE table_schema = 'public' 
    AND table_name LIKE '%ticket%'
    ORDER BY table_name;
  `);
  
  console.log('Tablas encontradas:', tablesResult.rows.map(r => r.table_name).join(', '));

  // 2. Analizar estructura de tickets
  console.log('\n📋 Estructura de tickets:');
  const ticketsColumns = await client.query(`
    SELECT column_name, data_type, is_nullable 
    FROM information_schema.columns 
    WHERE table_name = 'tickets' 
    ORDER BY ordinal_position;
  `);
  
  ticketsColumns.rows.forEach(col => {
    console.log(`  - ${col.column_name}: ${col.data_type} ${col.is_nullable === 'NO' ? 'NOT NULL' : ''}`);
  });

  // 3. Ver índices existentes en tickets
  console.log('\n📇 Índices existentes en tickets:');
  const ticketsIndexes = await client.query(`
    SELECT indexname, indexdef 
    FROM pg_indexes 
    WHERE tablename = 'tickets';
  `);
  
  ticketsIndexes.rows.forEach(idx => {
    console.log(`  - ${idx.indexname}`);
  });

  // 4. Ver estructura de ticket_costs
  console.log('\n💰 Estructura de ticket_costs:');
  const costsColumns = await client.query(`
    SELECT column_name, data_type, is_nullable 
    FROM information_schema.columns 
    WHERE table_name = 'ticket_costs' 
    ORDER BY ordinal_position;
  `);
  
  costsColumns.rows.forEach(col => {
    console.log(`  - ${col.column_name}: ${col.data_type} ${col.is_nullable === 'NO' ? 'NOT NULL' : ''}`);
  });

  // 5. Ver índices existentes en ticket_costs
  console.log('\n📇 Índices existentes en ticket_costs:');
  const costsIndexes = await client.query(`
    SELECT indexname, indexdef 
    FROM pg_indexes 
    WHERE tablename = 'ticket_costs';
  `);
  
  costsIndexes.rows.forEach(idx => {
    console.log(`  - ${idx.indexname}`);
  });

  // 6. Verificar funciones RPC existentes
  console.log('\n🔧 Funciones RPC (procedimientos):');
  const funcsResult = await client.query(`
    SELECT proname, prosrc 
    FROM pg_proc 
    WHERE proname LIKE '%ticket%' 
    AND pronamespace = (SELECT oid FROM pg_namespace WHERE nspname = 'public');
  `);
  
  funcsResult.rows.forEach(f => {
    console.log(`  - ${f.proname}`);
    // Mostrar preview del código
    if (f.prosrc && f.prosrc.length > 200) {
      console.log(`    Código: ${f.prosrc.substring(0, 200)}...`);
    } else {
      console.log(`    Código: ${f.prosrc}`);
    }
  });

  // 7. Ver tamaño de las tablas
  console.log('\n📦 Tamaño de tablas:');
  const sizeResult = await client.query(`
    SELECT 
      relname as table_name,
      pg_size_pretty(pg_total_relation_size(relid)) as size
    FROM pg_catalog.pg_statio_user_tables
    WHERE relname IN ('tickets', 'ticket_costs')
    ORDER BY pg_total_relation_size(relid) DESC;
  `);
  
  sizeResult.rows.forEach(r => {
    console.log(`  - ${r.table_name}: ${r.size}`);
  });

  // 8. Contar registros
  console.log('\n📊 Conteo de registros:');
  const countTickets = await client.query('SELECT COUNT(*) as count FROM tickets');
  console.log(`  - Tickets: ${countTickets.rows[0].count}`);
  
  try {
    const countCosts = await client.query('SELECT COUNT(*) as count FROM ticket_costs');
    console.log(`  - Ticket Costs: ${countCosts.rows[0].count}`);
  } catch (e) {
    console.log('  - Ticket Costs: tabla no existe o error');
  }

  // 9. Ver queries lentos recientes (si hay pg_stat_statements)
  console.log('\n🐢 Queries lentos (si disponibles):');
  try {
    const slowQueries = await client.query(`
      SELECT query, calls, mean_time, total_time
      FROM pg_stat_statements
      WHERE query LIKE '%ticket%'
      ORDER BY mean_time DESC
      LIMIT 5;
    `);
    
    if (slowQueries.rows.length > 0) {
      slowQueries.rows.forEach(q => {
        console.log(`  - Mean: ${Math.round(q.mean_time)}ms, Calls: ${q.calls}, Query: ${q.query.substring(0, 80)}...`);
      });
    } else {
      console.log('  No hay datos de queries lentos (pg_stat_statements no habilitado)');
    }
  } catch (e) {
    console.log('  pg_stat_statements no disponible');
  }

  await client.end();
  console.log('\n✅ Diagnóstico completado');
}

main().catch(console.error);